shift/src/autogluon/shift/learners/xShiftDetector.py

```diff
@@ -0,0 +1,154 @@
+
+    Parameters
+    ----------
+    PredictorClass: an AutoGluon predictor, such as instance of autogluon.tabular.Predictor
+        The predictor that will be fit on training set and predict the test set
+
+    label: str
+        the Y variable that is to be predicted (if it appears in the train/test data then it will be removed)
+
+    classification_metric: str
+        the metric used for the C2ST, it must be one of ['accuracy']
+
+    Methods
+    -------
+    fit: fits the detector on training and test covariate data
+
+    json, summary: outputs the results of XShift detection
+    - test statistic
+    - pvalue (optional, if compute_pvalue=True in .fit)
+    - detector feature importances
+    - top k anomalous samples
+
+    Usage
+    -----
+    >>> xshiftd = XShiftDetector(TabularPredictor, label='class')
+    Fit the detector...
+    >>> xshiftd.fit(Xtrain, Xtest)
+    Output the decision...
+    >>> xshiftd.decision()
+    Output the summary...
+    >>> xshiftd.summary()
+    """
+
+    def __init__(self,
+                 PredictorClass,
+                 label=None,
+                 classification_metric = 'balanced accuracy'):
+        named_metrics = {
+            'balanced accuracy' : balanced_accuracy_score,
+        }
+        assert classification_metric in named_metrics.keys(), \
+            'classification_metric must be one of [' + ', '.join(named_metrics.keys()) + ']'
+        pred = PredictorClass(label='xshift_label')
+        self.C2ST = sft.Classifier2ST(pred)
+        self.cmetric = named_metrics[classification_metric]
+        self.cmetric_name = classification_metric
+        if not label:
+            warnings.warn('label is not specified, please ensure that Xtrain, Xtest do not have the Y (label) '
+                          'variable')
+        self.label = label
+        self._is_fit = False
+
+    def _post_fit(func):
+        """decorator for post-fit methods"""
+        def pff_wrapper(self, *args, **kwargs):
+            assert self._is_fit, f'.fit needs to be called prior to .{func.__name__}'
+            return func(self, *args, **kwargs)
+        return pff_wrapper
+
+    def fit(self, Xtrain, Xtest):
+        """Fit the XShift detector.
+
+        Parameters
+        ----------
+        Xtrain, Xtest : pd.DataFrame
+            training dataframe and test dataframe
+        """
+        assert 'xshift_label' not in Xtrain.columns, 'your data columns contain "xshift_label" which is used internally'
+
+        if self.label:
+            Xtrain = Xtrain.drop(columns=[self.label])
+            Xtest = Xtest.drop(columns=[self.label])
+
+        self.C2ST.fit((Xtrain, Xtest), sample_label='xshift_label', accuracy_metric=self.cmetric)
+
+        ## Sample anomalies
+        as_top = self.C2ST.sample_anomaly_scores(how='top')
+        as_top = as_top[[1]].rename(columns={1: 'xshift_test_proba'})
+        self.anomalies = as_top.join(Xtest)
```

Comment:
This is of limited use without cross-val / out-of-fold pred proba because you are missing the anomalies used to train the predictor. 2-fold cross val would probably make this a lot nicer

Reply:
got it